pandas/tests/io/test_sql.py

```diff
@@ -563,12 +596,29 @@ def test_dataframe_to_sql_arrow_dtypes(conn, request):
                 [datetime(2023, 1, 1)], dtype="timestamp[ns][pyarrow]"
             ),
             "date": pd.array([date(2023, 1, 1)], dtype="date32[day][pyarrow]"),
-            "timedelta": pd.array([timedelta(1)], dtype="duration[ns][pyarrow]"),
             "string": pd.array(["a"], dtype="string[pyarrow]"),
         }
     )
+
+    if "adbc" in conn:
```

Comment:
@jorisvandenbossche we chatted about this at the dev meeting, but can't remember the resolution. The current postgres driver writes the integral value to postgres:

```sh
pandas=# select * FROM test_arrow;
 int |      datetime       |    date    | string |   timedelta    
-----+---------------------+------------+--------+----------------
   1 | 2023-01-01 00:00:00 | 2023-01-01 | a      | 86400000000000
(1 row)
```

The ADBC driver can write the timedelta as a postgresl interval:

```sh
pandas=# select * FROM test_arrow;
 int |      datetime       |    date    | string | timedelta 
-----+---------------------+------------+--------+-----------
   1 | 2023-01-01 00:00:00 | 2023-01-01 | a      | 1 day
(1 row)
```

IIRC from our conversation you thought this should stay duration and we should follow through with https://github.com/apache/arrow-adbc/pull/907 upstream. However, I intentionally closed that in favor of https://github.com/apache/arrow-adbc/pull/908 being of the opinion that a timedelta maps more closely to an interval than it does a pyarrow duration

Does that correctly summarize what we talked about? Do you still think this usage is incorrect?

Reply:
> However, I intentionally closed that in favor of [apache/arrow-adbc#908](https://github.com/apache/arrow-adbc/pull/908) being of the opinion that a timedelta maps more closely to an interval than it does a pyarrow duration

IIUC that merged PR added support for reading/writing Postgres' Interval <-> Arrow's Interval[MONTH_DAY_NANO], and didn't add anything to handle Arrow's duration type. That is correct?

So I think for Postgres<->Arrow interaction, that's certainly the most native, default conversion. 
But there is a separate question how to write Arrow's furation type to Postgres. Fail, store as integer, or store as interval with only microseconds set (no months/days) (and it's my understanding that you actually implemented this last option in the closed PR?)

Separately, for pandas, there is the question what the best mapping is for pandas<->arrow. For our timedelta64, that is definitely Arrow's duration (that is the conversion that is implemented by pyarrow for pandas<->arrow conversions). And then how we can support pandas' timedelta64 in `to_sql` based on adbc depends on the question above about how to handle duration in the adbc postgres driver.

Reply:
Yes that PR was only for interval. And you are correct - the other PR checked the bounds of the duration and raised if it was OOB for postgres' storage capabilities; if not it stored it as an interval. I am happy to re-open that PR and see it through if we see value in that.

I'm a bit confused by the choice of duration though. Don't both pandas and CPython use days and time as separate components? I see this in pandas:

```c
typedef struct {
        npy_int64 days;
        npy_int32 hrs, min, sec, ms, us, ns, seconds, microseconds, nanoseconds;
} pandas_timedeltastruct;
```

and this in CPython

```c
typedef struct
{
    PyObject_HEAD
    Py_hash_t hashcode;         /* -1 when unknown */
    int days;                   /* -MAX_DELTA_DAYS <= days <= MAX_DELTA_DAYS */
    int seconds;                /* 0 <= seconds < 24*3600 is invariant */
    int microseconds;           /* 0 <= microseconds < 1000000 is invariant */
} PyDateTime_Delta;
```

Which I think can both be losslessly converted to/from the month/day/nano struct but not to a duration that only stores milliseconds?

Reply:
> I'm a bit confused by the choice of duration though. Don't both pandas and CPython use days and time as separate components?

The numpy `timedelta64[unit]` dtype we use in pandas actually does store just int64 numbers, and thus it is exactly the same as the `duration[unit]` type in Arrow (minus the validity bitmap). See https://numpy.org/doc/stable/reference/arrays.datetime.html#datetime-units

The `pandas_timedeltastruct` you mention is used to convert this single integer number into the components (eg for when accessing one of those components), but so for a full column we don't store those components (for a scalar Timedelta object, I think the components might be cached once accessed).

For the stdlib `datetime.timedelta`, it is true this internally this stores components: days, seconds, microseconds (https://docs.python.org/3/library/datetime.html#timedelta-objects), and thus can be considered a "relative" timedelta. Although it's also a bit in between of an "absolute" or "relative" time delta. On the one hand, it considers 3600*24 seconds to be one day (when constructing it) and doesn't support storing months, and for example based on that rule, it can always unambiguously be converted to a single number for `pd.Timedelta` / numpy `timedelta64` / arrow `duration`. But on the other hand, in arithmetic operations, it then _does_ behave as a relative timedelta, and so you do see differences between stdlib timedelta and pd.Timedelta (although it's officially a subclass ..):

```python
import datetime
import zoneinfo

dt = datetime.datetime(2023, 10, 28, 9, tzinfo=zoneinfo.ZoneInfo("Europe/Brussels"))
print(dt)
# 2023-10-28 09:00:00+02:00
print(dt + datetime.timedelta(days=1))
# 2023-10-29 09:00:00+01:00   # <-- because of DST, the actual time difference is 25 hours, not 24 hours
print(dt + datetime.timedelta(seconds=24*3600))
# 2023-10-29 09:00:00+01:00   # <-- could be considered wrong because I asked for 24 hours but got 25?

# in contrast, using pandas Timestamp, we get a different result because 1 day == 24 hours
print(pd.Timestamp(dt) + pd.Timedelta("1 day"))
# 2023-10-29 08:00:00+01:00
print(pd.Timestamp(dt) + pd.Timedelta("24 hours"))
# 2023-10-29 08:00:00+01:00
```

So for the stdlib `datetime.timedelta`, I think it's correct to say this might map better to an arrow Interval (although it's limited to "days", for a more complete relative timedelta, there is `dateutil.relativedelta`, which also supports years and months). 
I suppose the main reason that we convert `datetime.timedelta` (somewhat relative) to `np.timedelta64` (absolute) is that historically, there hasn't really been any alternative. In pandas we use the numpy dtypes, and then the absolute timedelta64 was the only option. We do have our own DateOffsets that are meant to be relative, but those have never been completely integrated (for example not as an actual column dtype). 
While Arrow did have an Interval type from the beginning (although not the month_day_nano), I suppose pyarrow essentially copied pandas in converting datetime.timedelta to timedelta64/duration. The support for Interval types in pyarrow is also very limited (eg we didn't even expose those types in the bindings, until recently only the month_day_nano was added).

Now, for the original question what to do for ADBC: in pandas we already made the conversion from `datetime.timedelta` to `timedelta64`, and at that point this is essentially mapping to Arrow `duration`, and not Arrow `interval`. 
In theory, a user could store `datetime.timedelta` objects in a `object` dtype column (or use ArrowDtype with month_day_nano_interval), and then specify the interval type for the schema when converting pandas->arrow, as a way to insert Interval data in a postgres database. But that's not really ideal, but I don't think there is a way around that until pandas has proper relative timedelta (interval) support.

Reply:
Thanks for the details - very interesting. I can definitely add the duration support back upstream. Postgres, which AFAIK stores and manages everything in UTC, should therefore not have the ambiguity of a timedelta crossing DST boundaries as I guess it would always consider a day 24 hours. So I can pick back up the duration writing upstream; the big downside to that is postgres having an int64 limit on the microseconds portion, but I think that is a forward-looking concern for pandas that fits where we are today

With that said, I don't know if that is going to be a blocker? Writing timedeltas to a database probably isn't critical so I am hoping we don't need to block this as we wait for that

Reply:
Note that it does raise in the current solution. You'd have to explicitly convert it to `month_day_nano` for it to work in this PR